### PR TITLE
Added subscribe api

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
#### What are the relevant tickets?
fixes https://github.com/mitodl/open-discussions/issues/112

#### What's this PR do?
creates 2 apis one is subscribe and one is un subscribe from channel

#### How should this be manually tested?
- Create 2 channels
- Listing channels http://192.168.99.100:8063/api/v0/channels/
- Unsubscribe one channel http://192.168.99.100:8063/api/v0/channels/{:url channel name}/unsubscribe/
- List channels, You wont be able to see it unsub channel
- Subscribe channel http://192.168.99.100:8063/api/v0/channels/test_channel/subscribe/
- List channel, Now you can see it.
 
@pdpinch 

#### Screenshots (if appropriate)
Output
screen shot 2017-09-07 at 7 33 18 pm

### CODE REVIEW

File: channels/api.py [hunk 1]
```diff
@@ -361,3 +363,52 @@ def more_comments(self, comment_fullname, parent_fullname, count, children=None)
...
+            channels.models.UnSubscribeStatus: the subscription status of user
+        """
+        status, __ = UnSubscribeStatus.objects.get_or_create(user=self.user, channel_name=channel_name)
+        if not self.is_subscribe(channel_name):
+            self.get_channel(channel_name).subscribe()
+            status.is_unsubscribe = False
+            status.save()
+
+        return {
+            'is_subscribed': self.is_subscribe(channel_name)
```
Comment (amir-qayyum-khan) [inline channels/api.py]: it is out put is correct

File: channels/api.py [hunk 2]
```diff
@@ -361,3 +363,52 @@ def more_comments(self, comment_fullname, parent_fullname, count, children=None)
...
+        """
+        Unsubscribe a user to a channel
+
+        Args:
+            channel_name(str): the channel name identifier
+
+        Returns:
+            channels.models.UnSubscribeStatus: the subscription status of user
+        """
+        status, __ = UnSubscribeStatus.objects.get_or_create(user=self.user, channel_name=channel_name)
```
Comment (amir-qayyum-khan) [inline channels/api.py]: why we are tracking un subscription of users, if you unsubscribe what will happen then

File: channels/models.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+"""Subscribe models"""
+from django.db import models
+from django.conf import settings
+
+
+class UnSubscribeStatus(models.Model):
```
Comment (amir-qayyum-khan) [inline channels/models.py]: see https://github.com/mitodl/open-discussions/issues/112

File: channels/urls.py [hunk 4]
```diff
@@ -9,6 +9,8 @@
     FrontPageView,
     PostDetailView,
     PostListView,
+    SubscribeView,
```
Comment (amir-qayyum-khan) [inline channels/urls.py]: @rhysyngsun 2 end points are ok or should i made them one

File: channels/api.py [hunk 5]
```diff
@@ -361,3 +363,52 @@ def more_comments(self, comment_fullname, parent_fullname, count, children=None)
         more_comments.submission = self.reddit.submission(submission_id)
         more_comments.comments()  # load the comments
         return more_comments
+
```
Comment (amir-qayyum-khan) [inline channels/api.py]: @rhysyngsun please review this method

File: channels/api.py [hunk 6]
```diff
@@ -361,3 +363,52 @@ def more_comments(self, comment_fullname, parent_fullname, count, children=None)
...
+    def is_subscribe(self, channel_name):
```
Comment (amir-qayyum-khan) [inline channels/api.py]: @rhysyngsun please review this method

**Comment by noisecapella** [comment]
When you mark a PR 'Needs review' there should be tests already written for it

File: channels/urls.py [hunk 7]
```diff
@@ -9,6 +9,8 @@
...
+    UnSubscribView
```
Comment (noisecapella) [inline channels/urls.py]: Please change this to `UnsubscribeView`

File: channels/views.py [hunk 8]
```diff
@@ -105,3 +107,29 @@ def get_queryset(self):
...
+
+class UnSubscribView(RetrieveUpdateAPIView):
+    """
+    View to retrieve and remove subscriber in channels
+    """
+    permission_classes = (IsAuthenticated, JwtIsStaffOrReadonlyPermission, )
+    serializer_class = SubscriberSerializer
+
+    def get_object(self):
+        """unsub user"""
```
Comment (noisecapella) [inline channels/views.py]: No, this is not where we should do this. `get_object` is for retrieving the object matching the REST API.

**Comment by noisecapella** [comment]
It might be useful to look at @giocalitri's work on the contributors API: https://github.com/mitodl/open-discussions/pull/140/files

One thing to note especially: the REST API for contributors is `/api/v0/channels/<channel_name>/contributors/`. `POST` is used at this endpoint to add a contributor to the channel, and `DELETE` is used with the contributor detail view to remove a contributor. We should change the endpoint to `/api/v0/channels/<channel_name>/subscribers/` and put together the REST API similarly.

**Comment by pdpinch** [comment]
Amir, I'm going to reassign this since we want it be more like https://github.com/mitodl/open-discussions/pull/140/files and it will be faster if someone in the office does it.

**Pull request closed without merging**